### PR TITLE
predict sql submit predict job as maxcompute udf script

### LINKED ISSUE #604: SQLFlow Predict SQL generate ODPS UDF program
**Is your feature request related to a problem? Please describe.**

SQLFlow would use ODPS UDF as the Alps prediction Job, for a simple example prediction UDF program:

``` text
set mst.model.name=push_common.model;
set mst.oss.id=aaa;
set mst.oss.key=bbb;

.....

SELECT sqlflow_projet.alps_inference(
            concat_ws(",",user_id,label)
            ,1
            ,"deep_id"
            ,COALESCE(deep_id, "")
        ) as (user_info, score)
FROM sqlflow_projet.users;
```

The SQLFlow Predict SQL can be like:

```  SQL
SELECT sqlflow_projet.alps_inference(
            concat_ws(',',user_id,label)
            ,1
            ,"deep_id"
            ,COALESCE(deep_id, "")
        ) AS (user_info, score)
FROM sqlflow_projet.users
PREDICT sqlflow_project.pred_result
USING sqlflow_model;
```

There would be two tasks to finish this feature:
- [x] SELECT fields syntax need to support ODPS UDF syntax. #600 
- [x] codegen_alsp can generate the ODPS UDF program and execute through the ODPS database interface. #604 
- [ ] [gomaxcompute](https://github.com/sql-machine-learning/gomaxcompute) need to support adding the query  settings. https://github.com/sql-machine-learning/gomaxcompute/issues/45

NOTE: For the future, we need a more general solution to parse the standard SQL syntax.

### PR DESCRIPTION
**Pull request by Yancey0623**
Related issue: #604 

An example SQL can be like:

``` sql
SELECT alps_inference(
            concat_ws(",",user_id,l)
            ,1
            ,"deep_id"
            ,COALESCE(deep_id, "")
        ) AS (user_info, score)
FROM alifin_jtest_dev.predict_data
PREDICT predict_result
WITH OSS_ID=a, OSS_KEY=b
USING sqlflow_model;
```

TODO:  https://github.com/sql-machine-learning/gomaxcompute/pull/46

### CODE REVIEW

**Comment by Yancey0623** [comment]
Execute the ODPS query failed:
``` text
response error: 400, <?xml version="1.0" encoding="UTF-8"?>
<Error>
	<Code>ParseError</Code>
	<Message><![CDATA[Please add put { "odps.sql.submit.mode" : "script"} for multi-statement query in settings.]]></Message>
	<RequestId>5D3BE689B8C6891C0879BE45</RequestId>
	<HostId>odps.aliyun.com</HostId>
</Error>
```

~~the `gomaxcompute` need to support the `settings` interface.~~

This PR uses `odpscmd` to submit the ODPS UDF script.

File: scripts/image_build.sh [hunk 1]
```diff
@@ -101,7 +101,14 @@ if [ "${WITH_SQLFLOW_MODELS:-ON}" = "ON" ]; then
   rm -rf models
 fi
 
-# 7. Load sqlflow Jupyter magic command automatically. c.f. https://stackoverflow.com/a/32683001.
+# 7. Install odpscmd for submitting alps predict job with odps udf script
+# TODO(Yancey1989): using gomaxcompute instead of the odpscmd command-line tool.
+wget -q http://docs-aliyun.cn-hangzhou.oss.aliyun-inc.com/assets/attach/119096/cn_zh/1557995455961/odpscmd_public.zip
```
Comment (typhoonzero) [inline scripts/image_build.sh:104]: Make sure `odpscmd` a public available program.

File: sql/codegen_alps.go [hunk 2]
```diff
@@ -53,13 +57,22 @@ type alpsFiller struct {
 	TrainClause       *resolvedTrainClause
 	ExitOnSubmit      bool
 
+	// Predict
+	PredictStanderClause string
```
Comment (typhoonzero) [inline sql/codegen_alps.go]: standard?

File: sql/codegen_alps.go [hunk 3]
```diff
@@ -40,6 +43,7 @@ type alpsFiller struct {
 	ModelDir           string
 	ScratchDir         string
 	PredictOutputTable string
+	PredictInputModel  string
```
Comment (typhoonzero) [inline sql/codegen_alps.go]: Can reuse `ModelDir` ?

File: sql/sql.y [hunk 4]
```diff
@@ -175,18 +176,17 @@ select
...
   }
 ;
 
 predict_clause
-: PREDICT IDENT USING IDENT {
-	$$.into = $2
-	$$.model = $4
-}
+: PREDICT IDENT { $$.into = $2 }
+| PREDICT WITH attrs { $$.predAttrs = $3 }
```
Comment (typhoonzero) [inline sql/sql.y]: curious why not `PREDICT IDENT  WITH attrs` ?

File: sql/sql.y [hunk 5]
```diff
@@ -175,18 +176,17 @@ select
...
+| predict_clause USING IDENT { $$.model = $3 }
```
Comment (typhoonzero) [inline sql/sql.y]: Does this line mean we can write multiple `USING`?

File: scripts/image_build.sh (hunk 1 above)
Comment (Yancey0623) [reply scripts/image_build.sh:104]: I copied the link from Aliyun, I think it's a public program.

File: sql/sql.y (hunk 4 above)
Comment (Yancey0623) [reply sql/sql.y]: Fixed it.

File: sql/codegen_alps.go (hunk 2 above)
Comment (Yancey0623) [reply sql/codegen_alps.go]: done.

File: sql/codegen_alps.go (hunk 3 above)
Comment (Yancey0623) [reply sql/codegen_alps.go]: Yes, fixed it.

File: sql/sql.y (hunk 5 above)
Comment (Yancey0623) [reply sql/sql.y]: Done, and found some same syntax bug, such as `SELECT xx FROM a FROM b`, `SELECT xxx from table TRAIN model COLUMN a,b COLUMN c,d`, I issued a bug: https://github.com/sql-machine-learning/sqlflow/issues/616 , and will fix it in another pr.

**Review by tonyyang-svail — DISMISSED** [review]
Generally looks good to me.

File: sql/codegen_alps.go [hunk 6]
```diff
@@ -255,13 +267,37 @@ func newALPSTrainFiller(pr *extendedSelect, db *DB, session *pb.Session, ds *tra
...
+	if ossID, ok = pr.predAttrs["OSS_ID"]; !ok {
+		return nil, fmt.Errorf("the ALPS Predict job should specify OSS_ID")
+	}
+	if ossKey, ok = pr.predAttrs["OSS_KEY"]; !ok {
+		return nil, fmt.Errorf("the ALPS Predict job should specify OSS_KEY")
+	}
+	modelDir := fmt.Sprintf("oss://arks-model/%s/%s.tar.gz", session.UserId, pr.predictClause.model)
+
+	return &alpsFiller{
+		IsTraining:         true,
```
Comment (typhoonzero) [inline sql/codegen_alps.go]: Why predict should set `IsTraining:         true,`?

File: sql/codegen_alps.go [hunk 7]
```diff
@@ -293,24 +329,53 @@ pip install http://091349.oss-cn-hangzhou-zmf.aliyuncs.com/alps/sqlflow/alps-2.0
...
-	return submitALPS(w, cwd, f)
-}
+	if err = alpsPredTemplate.Execute(&program, filler); err != nil {
+		return fmt.Errorf("submitALPS: failed executing template: %v", err)
+	}
 
-func alpsPred(w *PipeWriter, pr *extendedSelect, db *DB, cwd string, session *pb.Session) error {
-	f, err := newALPSPredictFiller(pr)
+	fname := "alps_pre.odps"
+	filepath := filepath.Join(cwd, fname)
```
Comment (typhoonzero) [inline sql/codegen_alps.go:347]: defer clean up this file?

File: sql/codegen_alps.go (hunk 6 above)
Comment (Yancey0623) [reply sql/codegen_alps.go]: done.

File: sql/codegen_alps.go (hunk 7 above)
Comment (Yancey0623) [reply sql/codegen_alps.go:347]: done.

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**